File: Sem1-2/CPP_Java2024/List3/JAVA/Rectangle.java
public class Rectangle extends Quad
{
    protected final double side_1;
    protected final double side_2;

    public Rectangle(double side_1, double side_2) throws IllegalArgumentException
    {
        super("Rectangle");
        if(side_1 < 0 || side_2 < 0)
        {
            throw new IllegalArgumentException("Rectanle sides cannot be negative, got: " + side_1 + " " + side_2);
        }
        this.side_1 = side_1;
        this.side_2 = side_2;
    }

    @Override
    public double Circumference()
    {
        return 2.0 * side_1 + 2.0 * side_2;
    }

    @Override
    public double Area()
    {
        return side_1 * side_2;
    }
}

File: Sem1-2/CPP_Java2024/List3/CPP/src/Rectangle.cpp
#include <Rectangle.hpp>
#include <utils.hpp>

#include <stdexcept>

Rectangle::Rectangle(double side_1, double side_2) : Quad("Rectangle"), side_1(side_1), side_2(side_2)
{
    if(side_1 < 0 || side_2 < 0)
    {
        throw std::invalid_argument(LOG_LOC() + "Rectangle sides cannot be negative, got: " + std::to_string(side_1) + " " + std::to_string(side_2));
    }
}

double Rectangle::Circumference() const
{
    return 2.0 * side_1 + 2.0 * side_2;
}

double Rectangle::Area() const
{
    return side_1 * side_2;
}